fondant/compiler.py

```diff
@@ -0,0 +1,107 @@
+import json
+import logging
+import typing as t
+from abc import ABC, abstractmethod
+from pathlib import Path
+
+import yaml
+
+from fondant.pipeline import ComponentOp, Pipeline
+
+logger = logging.getLogger(__name__)
+
+
+class Compiler(ABC):
+    """Abstract base class for a compiler."""
+
+    def __init__(self, pipeline: Pipeline):
+        self.pipeline = pipeline
+
+    @abstractmethod
+    def compile(self):
+        """Abstract method to invoke compilation."""
+
+
+class DockerCompiler(Compiler):
+    """Compiler that creates a docker-compose spec from a pipeline."""
+
+    def compile(self, package_path: str = "docker-compose.yml") -> None:
+        """Compile a pipeline to docker-compose spec that is saved on the package_path."""
+        logger.info(f"Compiling {self.pipeline.name} to docker-compose.yml")
+        self._patch_path()
+        spec = self._generate_spec()
+        with open(package_path, "w") as outfile:
+            yaml.safe_dump(spec, outfile)
+        logger.info(f"Successfully compiled to {package_path}")
+
+    @staticmethod
+    def _safe_component_name(component_name: str) -> str:
+        """Transform a component name to a docker-compose friendly one.
+        eg: `Component A` -> `component_a`.
+        """
+        return component_name.replace(" ", "_").lower()
+
+    def _patch_path(self):
+        """Helper that checks if the base_path is local or remote,
+        if local it patches the base_path and prepares a bind mount.
+        """
+        base_path = Path(self.pipeline.base_path)
+        # check if base path is an existing local folder
+        if base_path.exists():
+            self.volume = {
+                "type": "bind",
+                "source": str(base_path),
+                "target": f"/{base_path.stem}",
+            }
+            self.path = f"/{base_path.stem}"
```

Comment:
Pylint would throw an `attribute-defined-outside-init` here :)

It's a bit hard to follow currently. Can we move the `_patch_path()` call to the `_generate_spec` method and pass them to `compose_service`? This is the only place they're used.

So something like:

```Python
def _generate_spec(self) -> dict:
    ...
    path, volume = self._patch_path()
    for component_name, component in self.pipeline._graph.items():
        ...
        services[safe_component_name] = self.compose_service(
            component["fondant_component_op"], 
            dependencies=component["dependencies"],
            path=path,
            volume=volume,
        )
```